sdk/src/main/scala/com/akkaserverless/javasdk/impl/eventsourcedentity/EventSourcedEntitiesImpl.scala

```diff
@@ -177,53 +183,84 @@ final class EventSourcedEntitiesImpl(_system: ActorSystem,
                                    service.snapshotEvery,
                                    log)
 
-          val reply: Reply[JavaPbAny] = try {
-            handler.handleCommand(cmd, context)
+          // FIXME we'd want to somehow share this handle-command-apply-event logic to get the end effect ready for asserting in the testkit
+          // FIXME a bit mixed concerns here, esp with the serialization to PbAny but it's either that or pushing this into the handler and making
+          // SecondaryEffectImpl a public API (or make handler internal, which may be a good idea, also for the testkit)
+          var endSequenceNumber = sequence
+          val (events: Vector[ScalaPbAny], secondaryEffect: SecondaryEffectImpl, snapshot: Option[ScalaPbAny]) = try {
+            val commandEffect =
+              handler.handleCommand(cmd, context).asInstanceOf[EventSourcedEntityEffectImpl[Any]]
+            commandEffect.primaryEffect match {
+              case EmitEvents(events) =>
+                var shouldSnapshot = false
+                val scalaPbEvents = Vector.newBuilder[ScalaPbAny]
+                events.foreach { event =>
+                  val javaPbEvent = service.anySupport.encodeJava(event)
+                  scalaPbEvents += ScalaPbAny.fromJavaProto(javaPbEvent)
+                  endSequenceNumber = endSequenceNumber + 1
+                  handler.handleEvent(javaPbEvent, new EventContextImpl(thisEntityId, sequence))
+                  shouldSnapshot = shouldSnapshot || (service.snapshotEvery > 0 && endSequenceNumber % service.snapshotEvery == 0)
+                }
+                // FIXME currently snapshotting final state after applying all events even if trigger was mid-event stream?
+                val snapshot =
+                  if (shouldSnapshot) Option(ScalaPbAny.fromJavaProto(handler.currentState()))
+                  else None
+                (scalaPbEvents.result(),
+                 commandEffect.secondaryEffect(service.anySupport.decode(handler.currentState())),
+                 snapshot)
+              case NoPrimaryEffect =>
+                (Vector.empty, commandEffect.secondaryEffect(handler.currentState()), None)
+            }
```

Comment:
This is the interpretation "engine" that we might want to share with the testkit (but sans the back and forth with `PbAny`s) - invoke command handler, apply emitted events, figure out end state, potential snapshot and side effects (such as reply to the user)